clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java

```diff
@@ -129,14 +141,30 @@ public Metrics(MetricConfig defaultConfig, List<MetricsReporter> reporters, Time
      * @param enableExpiration true if the metrics instance can garbage collect inactive sensors, false otherwise
      */
     public Metrics(MetricConfig defaultConfig, List<MetricsReporter> reporters, Time time, boolean enableExpiration) {
```

Comment:
we seem to call this method without a context in KafkaServer. Instead, KafkaServer should provide an initial context with the correct namespace, and then do a subsequent call to contextChange once the additional fields are added.

Reply:
I updated KafkaServer.scala, using Metrics constructor with MetricsContext.

Reply:
But I can only pass JmxReporter to Metrics constructor, not KafkaMetricReporter.  https://github.com/xiaodongdu/kafka/blob/9511de85dddfd619ebb2a65156674c7aca399bba/core/src/main/scala/kafka/server/KafkaServer.scala#L253

In https://github.com/xiaodongdu/kafka/blob/9511de85dddfd619ebb2a65156674c7aca399bba/core/src/main/scala/kafka/server/DynamicBrokerConfig.scala#L796
it will call notifyMetricsReporter, which will call contextChange on MetricsReporter loaded from config file.

Reply:
I'm not sure what your concern is about `KafkaMetricReporter`, it was never passed to `new Metrics(...)` in the first place.